src/lib/core/func_adapter.h

```diff
@@ -0,0 +1,284 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct tuple;
+
+#define FUNC_ADAPTER_CTX_SIZE 24
+
+/**
+ * Abstract func_adapter_ctx instance. It is supposed to be converted
+ * to a concrete realization, which must not occupy more memory than this
+ * instance.
+ */
+struct func_adapter_ctx {
+	/**
+	 * Padding to achieve required size.
+	 */
+	char pad[FUNC_ADAPTER_CTX_SIZE];
+};
+
+struct func_adapter;
+
+/**
+ * Virtual table for func_adapter class.
+ * Usage of these methods is in the description of func_adapter class.
+ */
+struct func_adapter_vtab {
+	/**
+	 * Prepares for call of the function. One must allocate func_adapter_ctx
+	 * (using stack or heap) and pass it to this function to initialize it.
+	 * After this call, all the arguments must be pushed in direct order.
+	 */
+	void (*begin)(struct func_adapter *func, struct func_adapter_ctx *ctx);
+	/**
+	 * Calls the function. All the arguments must be pushed before.
+	 */
+	int (*call)(struct func_adapter_ctx *ctx);
+	/**
+	 * Releases all the resources occupied by context. It must not be used
+	 * after this method was called, so all required arguments must be
+	 * popped before.
+	 * Must be called even in the case of fail.
+	 */
+	void (*end)(struct func_adapter_ctx *ctx);
+	/**
+	 * Pushes tuple argument.
+	 */
+	void (*push_tuple)(struct func_adapter_ctx *ctx, struct tuple *tuple);
+	/**
+	 * Pushes number argument.
+	 */
+	void (*push_number)(struct func_adapter_ctx *ctx, double number);
+	/**
+	 * Pushes integer argument.
+	 */
+	void (*push_integer)(struct func_adapter_ctx *ctx, int64_t integer);
+	/**
+	 * Pushes string argument.
+	 */
+	void (*push_string)(struct func_adapter_ctx *ctx, const char *str,
+			    size_t len);
+	/**
+	 * Pushes null argument.
+	 */
+	void (*push_null)(struct func_adapter_ctx *ctx);
+	/**
+	 * Check if the next returned value is a tuple.
+	 */
+	bool (*is_tuple)(struct func_adapter_ctx *ctx);
+	/**
+	 * Pops tuple. Returned tuple is referenced by the function and caller
+	 * must unreference it. Never returns NULL.
+	 */
+	void (*pop_tuple)(struct func_adapter_ctx *ctx, struct tuple **tuple);
+	/**
+	 * Check if the next returned value is a number.
+	 */
+	bool (*is_number)(struct func_adapter_ctx *ctx);
```

Comment:
Does it return true if the value is of type `int64_t`? If yes, what would `pop_number` return in this case? Can it result in loss of precision? (e.g. 2^63-1 can't be stored in double precisely). Similar question to `is_integer`: does it return true if the value has a fractional part?

Please describe the behavior in the comment.

Reply:
Left such a comment:
```c
/**
  * Checks if the next returned value is a number that can be represented
  * by double without loss of precision.
  */
bool (*is_double)(struct func_adapter_ctx *ctx);
```